### PR TITLE
Move -epmd_module and -erl_epmd_port into kernel parameters

### PR DESCRIPTION
**Pull request by jonatanklosko**
This introduces `epmd_module` and `erl_epmd_listen_port` parameters for the kernel application. I updated the docs to deprecate `-epmd_module` and `-erl_epmd_port`, similarly to `-connect_all` in https://github.com/erlang/otp/commit/1d0fe6f71c320d2d1d19d9efd6dcba7b18bedca6.

The kernel application already has a lot of parameters related to distribution, so I believe semantically this makes sense.

For context, in [Livebook](https://github.com/livebook-dev/livebook) we started using a custom `epmd_module`. In Escript and Elixir releases, there are ways to set the `-epmd_module` flag, however there is no way to do that in development, other than setting `ERL_AFLAGS` env var. We already start distribution manually using `net_kernel:start/2`, so the idea is that now we could do `application:set_env(kernel, epmd_module, Module)` right before that.

### CODE REVIEW

File: lib/kernel/src/erl_epmd.erl [hunk 1]
```diff
@@ -164,14 +164,9 @@ to when accepting new distribution requests.
       Host :: atom() | string() | inet:ip_address(),
       Port :: non_neg_integer().
 listen_port_please(_Name, _Host) ->
-    try
-        %% Should come up with a new name for this as ERL_EPMD_PORT describes what
-        %% port epmd runs on which could easily be confused with this.
```
Comment (jonatanklosko) [inline lib/kernel/src/erl_epmd.erl:169]: Following this comment, I thought it's the perfect time to pick a different name, that's why I named the kernel parameter `erl_epmd_listen_port`.

File: lib/observer/src/observer_wx.erl [hunk 2]
```diff
@@ -776,12 +776,13 @@ get_nodes() ->
...
 erl_dist_port() ->
-    try
-        erl_epmd = net_kernel:epmd_module(),
-        {ok, [[StringPort]]} = init:get_argument(erl_epmd_port),
-        list_to_integer(StringPort)
-    catch
-        _:_ ->
+    case net_kernel:epmd_module() of
+        erl_epmd ->
+            case erl_epmd:listen_port_please(nonode, nohost) of
```
Comment (jonatanklosko) [inline lib/observer/src/observer_wx.erl:781]: Note the fake `nonode`, `nohost`, those arguments are ignored by `erl_epmd:listen_port_please/2`, and this way we use the public API. Alternatively we could expose `erl_epmd:erl_epmd_listen_port/0` and make it `-doc false`.

File: erts/doc/references/erl_cmd.md [hunk 3]
```diff
@@ -252,14 +252,13 @@ described in the corresponding application documentation.
   In this example, an Erlang runtime system is started with environment variable
   `DISPLAY` set to `gin:0`.
 
-- **`-epmd_module Module` (init flag)** - Configures the module responsible to
```
Comment (jonatanklosko) [inline erts/doc/references/erl_cmd.md:255]: Sidenote: I think neither of these is actually "(init flag)" at this point, because they are not interpreted by `init`.

File: lib/kernel/doc/kernel_app.md [hunk 4]
```diff
@@ -130,6 +130,26 @@ For more information about configuration parameters, see file
...
+- **`epmd_module = module()`{: #epmd_module }** - Configures the module
+  responsible for communication with [epmd](epmd_cmd.md). If this parameter
+  is undefined, it defaults to `erl_epmd`.
+
+  The now deprecated command line argument
+  [`-epmd_module <module>`](`e:erts:erl_cmd.md#epmd_module`) has the same
+  effect as the `epmd_module` configuration parameter. If this configuration
+  parameter is defined, it will override the command line argument.
+
+- **`erl_epmd_listen_port = integer()`** - Configures the port used by `m:erl_epmd`
```
Comment (josevalim) [inline lib/kernel/doc/kernel_app.md:142]: Perhaps we don't need the erl prefix?

File: lib/observer/src/observer_wx.erl (hunk 2 above)
Comment (josevalim) [reply lib/observer/src/observer_wx.erl:781]: We could also expose it as `net_kernel:epmd_listen_port`, to mirror `net_kernel:epmd_module`.

File: lib/kernel/doc/kernel_app.md (hunk 4 above)
Comment (jonatanklosko) [reply lib/kernel/doc/kernel_app.md:142]: That was my initial thought, but my reasoning is that this config option is only applicable when `epmd_module = erl_epmd`, in other words specific to `erl_epmd`. Does that makes sense?

File: lib/observer/src/observer_wx.erl (hunk 2 above)
Comment (jonatanklosko) [reply lib/observer/src/observer_wx.erl:781]: `epmd_module` is a broader config, while `erl_epmd_listen_port` is specific to `erl_epmd`, as per the other comment.

**Comment by jonatanklosko** [comment]
A follow up question, if this change is accepted, would it be possible to have it in the next maintenance release? If yes, I will change the target branch, and the I can submit a separate PR altering the `-erl_epmd_port 0` behaviour (if you agree that it's desired).

**Comment by jhogberg** [comment]
Thanks for the PR! I think we'll want to discuss this a bit more internally before merging, and most of us are on vacation right now. We'll revisit this later in the summer/early autumn. :-)

**Comment by jonatanklosko** [comment]
@jhogberg sounds good, thanks for the heads up! :)

File: lib/kernel/src/erl_epmd.erl [hunk 5]
```diff
@@ -291,26 +286,26 @@ init(_) ->
...
+		                    name = Name,
+		                    family = Family},
+                    {reply, {ok, Creation}, S};
                 Error ->
-                    case init:get_argument(erl_epmd_port) of
+                    case erl_epmd_listen_port() of
                         {ok, _} ->
                             {reply, {ok, -1}, State#state{ socket = -1,
                                                            port_no = PortNo,
                                                            name = Name} };
```
[the remark this answers is not in the document]
Comment (kikofernandez) [reply lib/kernel/src/erl_epmd.erl:302]: Making `Port > 0` is an incompatible change. What is the reason for checking that `Port > 0`?

File: lib/kernel/src/erl_epmd.erl (hunk 1 above)
Comment (kikofernandez) [reply lib/kernel/src/erl_epmd.erl:169]: `erl_epmd_node_listen_port` because this is only relevant to `erl_epmd`, but it is not the `listen_port` of `erl_epmd`, but of the node.

File: erts/preloaded/src/init.erl [hunk 6]
```diff
@@ -90,9 +90,6 @@ The `init` module interprets the following command-line flags:
   It defaults to `strict` from OTP 27 and this option is scheduled for removal
   in OTP 28.
 
-- **`-epmd_module Module`** - Specifies the module to use for registration and
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply erts/preloaded/src/init.erl:93]: Sure, I will mirror the docs from `erl_cmd.md`! My motivation for removing it here was that this flag doesn't really configure `init` and I assumed it is a leftover in the docs.

File: lib/kernel/src/erl_epmd.erl (hunk 1 above)
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl:169]: Sounds good!

File: lib/kernel/src/erl_epmd.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl:302]: Yes, since you are ok with moving it to `maint`, I won't change this in this PR!

---

I can still send a separate PR against `main` to check `Port > 0`, if you agree it makes sense.

Let me rephrase the above reasoning. This conditional says: it doesn't matter if epmd is running as long as `erl_epmd_listen_port` is set, because in that case we know which node port to talk to. However, if the user sets `erl_epmd_listen_port` to `0`, then we actually don't know the node port. Therefore, the proposed change is to only ignore epmd when the port is non-zero.

And to give another reason, the default behaviour of `erl_epmd` is as if `erl_epmd_listen_port = 0`, but in this code path setting `erl_epmd_listen_port = 0` explicitly does make a difference.

File: lib/kernel/src/erl_epmd.erl [hunk 7]
```diff
@@ -378,7 +373,24 @@ get_epmd_port() ->
 	error ->
 	    ?erlang_daemon_port
     end.
-	    
+
+erl_epmd_listen_port() ->
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl:377]: This change does not affect compatibility, so specifying either should work exactly the same. Or am I missing something else?

File: lib/kernel/src/net_kernel.erl [hunk 8]
```diff
@@ -2423,11 +2423,21 @@ protocol_childspecs([H|T]) ->
...
-	_ ->
-	    erl_epmd
+    case application:get_env(kernel, epmd_module) of
+      {ok, Module} when is_atom(Module) ->
+          Module;
+      {ok, Invalid} ->
+          error({invalid_parameter_value, epmd_module, Invalid});
+      undefined ->
+          Module =
+              case init:get_argument(epmd_module) of
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/net_kernel.erl:2433]: Sorry, which deprecation list do you mean? The changes in kernel notes.md?

File: lib/kernel/src/erl_epmd.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (kikofernandez) [reply lib/kernel/src/erl_epmd.erl]: Yes, specifying either should work, but specifying both of them should crash :)

File: lib/kernel/src/net_kernel.erl (hunk 8 above)
[the remark this answers is not in the document]
Comment (kikofernandez) [reply lib/kernel/src/net_kernel.erl:2433]: Sure, sorry for not being concrete.

I mean `$ERL_TOP/system/doc/general_info/DEPRECATIONS`.
The instructions for deprecating something are in the [DEPRECATE file](https://github.com/erlang/otp/blob/master/HOWTO/DEPRECATE.md) :)

File: lib/kernel/src/erl_epmd.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl]: > If version OTP 27.X is not compatible with OTP 27.Y, the user should choose the old one to maintain compatibility.

I think my confusion is because in this case the compatibility between version is kept, so it does not matter which option the user uses, the behaviour is the same.

File: lib/kernel/src/net_kernel.erl (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/net_kernel.erl:2433]: Perfect, thank you!

[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/net_kernel.erl:2433]: As far as I can tell, this file only includes function deprecations in a specific format, is there a room for erl flags? Or should it go in the `$ERL_TOP/system/doc/general_info/deprecations_27.md` log?

[the remark this answers is not in the document]
Comment (jhogberg) [reply lib/kernel/src/net_kernel.erl:2433]: > Or should it go in the `$ERL_TOP/system/doc/general_info/deprecations_27.md` log?

Yes.

File: lib/kernel/src/erl_epmd.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl]: So the idea is to do this?

```erlang
erl_epmd_node_listen_port() ->
    PortParameter =
        case application:get_env(kernel, erl_epmd_node_listen_port) of
            {ok, Port} when is_integer(Port), Port >= 0 ->
                {ok, Port};
            {ok, Invalid} ->
                error({invalid_parameter_value, erl_epmd_node_listen_port, Invalid});
            undefined ->
                undefined
        end,
    PortArgument =
        try
            {ok, [[StringPort]]} = init:get_argument(erl_epmd_port),
            IntPort = list_to_integer(StringPort),
            {ok, IntPort}
        catch error:_ ->
            undefined
        end,
    case {PortParameter, PortArgument} of
        {undefined, undefined} -> undefined;
        {_, undefined} -> PortParameter;
        {undefined, _} -> PortArgument;
        _ -> error({invalid_configuration, "either -erl_epmd_port or kernel erl_epmd_node_listen_port should be specified, not both"})
    end.
```

One downside of doing this is that we read both places on every call. The current implementation parses the argument on the first call and puts it in the kernel configuration, as done for `-connect_all` in https://github.com/erlang/otp/commit/1d0fe6f71c320d2d1d19d9efd6dcba7b18bedca6. WDYT?

[the remark this answers is not in the document]
Comment (kikofernandez) [reply lib/kernel/src/erl_epmd.erl]: I will get back to you on this one. I agree that it is not nice to read twice the same thing, and we also want to avoid having conflicting arguments, which is the reason for the crash

File: lib/kernel/doc/kernel_app.md [hunk 9]
```diff
@@ -130,6 +130,26 @@ For more information about configuration parameters, see file
     node. If a node goes down, it must thereafter be explicitly connected. See
     `m:net_kernel`.
 
+- **`epmd_module = module()`{: #epmd_module }** - Configures the module
+  responsible for communication with [epmd](epmd_cmd.md). If this parameter
```
Comment (garazdawi) [inline lib/kernel/doc/kernel_app.md:134]:
```suggestion
  responsible for communication with [epmd](`e:erts:erl_cmd.md`). If this parameter
```

Comment (garazdawi) [reply lib/kernel/doc/kernel_app.md]: actually I did a copy paste error in my fix proposal above. It should of course link to `epmd_cmd.md` and not `erl_cmd.md`.

File: lib/kernel/doc/kernel_app.md [hunk 10]
```diff
@@ -130,6 +130,26 @@ For more information about configuration parameters, see file
     node. If a node goes down, it must thereafter be explicitly connected. See
     `m:net_kernel`.
 
+- **`epmd_module = module()`{: #epmd_module }** - Configures the module
+  responsible for communication with [epmd](`e:erts:erl_cmd.md`). If this parameter
```
Comment (jonatanklosko) [inline lib/kernel/doc/kernel_app.md:134]:
```suggestion
  responsible for communication with [epmd](`e:erts:epmd_cmd.md`). If this parameter
```

File: lib/kernel/doc/kernel_app.md (hunk 9 above)
Comment (jonatanklosko) [reply lib/kernel/doc/kernel_app.md]: And when checking I looked at the links below, so I didn't notice. Updated!

File: lib/kernel/src/erl_epmd.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (kikofernandez) [reply lib/kernel/src/erl_epmd.erl]: Please, read it in both places even if it is cumbersome.
We would prefer a nice crash when there are conflicting options

Sorry for the delay, I forgot that I had to check this.

[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/kernel/src/erl_epmd.erl]: Updated :)

**Comment by kikofernandez** [comment]
If there is nothing left to fix, I will merge this next Monday (some of us are in a conference this week and I do not want to rush the merging before having a really final good look at it :) )

**Comment by jonatanklosko** [comment]
@kikofernandez sounds good, thanks :)

**Comment by jonatanklosko** [comment]
@kikofernandez kindly ping :)

**Comment by kikofernandez** [comment]
We have a code freeze for 27.1. I will merge as soon as I am given the OK.
This change was not breaking backwards compatibility, so we can merge in `maint` and release with OTP 27.2

**Comment by jonatanklosko** [comment]
@kikofernandez thank you :)
